bpf/bpf_lb.c

```diff
@@ -31,153 +31,109 @@
+			return ret;
 
-		}
+	}
 #endif
 
-		csum_off = l4_checksum_offset(nexthdr);
-		if (unlikely(!csum_off))
-			goto error;
-
-		switch (nexthdr) {
-		case IPPROTO_TCP:
-		case IPPROTO_UDP:
-			/* Port offsets for UDP and TCP are the same */
-			ret = l4_load_port(skb, l4_off + TCP_DPORT_OFF, &key.dport);
-			if (IS_ERR(ret))
-				goto error;
-
-			ret = l4_load_port(skb, l4_off + TCP_SPORT_OFF, &sport);
-			if (IS_ERR(ret))
-				goto error;
-			break;
-		/* FIXME: ICMPv6 */
-		default:
-			ret = DROP_UNKNOWN_L4;
-			goto error;
-		}
-
-		hash = lb_hash(ip6, sport, key.dport);
-		cilium_trace(skb, DBG_PKT_HASH, hash, 0);
-
-		key.dport = ntohs(key.dport);
-		//cilium_trace(skb, DBG_GENERIC, key.vip.p1, 0);
-		//cilium_trace(skb, DBG_GENERIC, key.vip.p2, 0);
-		//cilium_trace(skb, DBG_GENERIC, key.vip.p3, 0);
-		//cilium_trace(skb, DBG_GENERIC, key.vip.p4, 0);
-		//cilium_trace(skb, DBG_GENERIC, key.dport, 0);
-		val = map_lookup_elem(&cilium_lb_services, &key);
-		if (val != NULL && val->lxc_count) {
-			union macaddr lb_mac = NODE_MAC;
-			int i, which = hash % val->lxc_count;
-
-#pragma unroll
-			for (i = 0; i < MAX_LXC; i++) {
-				if (i != which)
-					continue;
-
-				ipv6_addr_copy(&lb_ip,
-					       (union v6addr *) &server_prefix);
-				ipv6_set_node_id(&lb_ip, val->lxc[i].node_id);
-				ipv6_set_state(&lb_ip, val->state);
-				ipv6_set_lxc_id(&lb_ip, val->lxc[i].lxc_id);
-				if (key.dport != val->lxc[i].port) {
-					__u16 tmp = htons(val->lxc[i].port);
-					//cilium_trace(skb, DBG_GENERIC, val->lxc[i].port, 0);
-					switch (nexthdr) {
-					case IPPROTO_TCP:
-					case IPPROTO_UDP:
-					/* Port offsets for UDP and TCP are the same */
-						ret = l4_modify_port(skb, l4_off + TCP_DPORT_OFF,
-								     csum_off, tmp,
-								     htons(key.dport));
-						if (IS_ERR(ret))
-							goto error;
-						break;
-					/* FIXME: Handle ICMPv6 */
-					/* default handled outside the loop */
-					}
-				}
-			}
-
-			ipv6_store_daddr(skb, lb_ip.addr, ETH_HLEN);
-
-			/* fixup csums */
-			sum = csum_diff(key.vip.addr, 16, lb_ip.addr, 16, 0);
-			if (l4_csum_replace(skb, csum_off, 0, sum, BPF_F_PSEUDO_HDR) < 0) {
-				ret = DROP_CSUM_L4;
-				goto error;
-			}
-
-			eth_store_saddr(skb, lb_mac.addr, 0);
-
-			/* Send the packet to the stack */
-			cilium_trace_capture(skb, DBG_CAPTURE_DELIVERY, 0);
-			skb->cb[CB_IFINDEX] = 0;
-			return TC_ACT_OK;
-		} else {
-			cilium_trace(skb, DBG_LB_SERVICES_LOOKUP_FAIL, key.vip.p4, key.dport);
-		}
+	ret = lb6_extract_key(skb, ip6, nexthdr, l4_off, &key, &csum_off, &csum_flags);
+	if (IS_ERR(ret))
+		return ret;
+
+	svc = lb6_lookup_service(skb, &key);
+	if (svc == NULL) {
+		/* Pass packets to the stack which should not be loadbalanced */
+		return TC_ACT_OK;
 	}
 
-	if (IS_ERR(ret)) {
-error:
+	hash = get_hash_recalc(skb);
+	slave = (hash % svc->count) + 1;
+	cilium_trace(skb, DBG_PKT_HASH, hash, slave);
+
+	if (!(svc = lb6_lookup_slave(skb, &key, slave)))
+		return DROP_NO_SERVICE;
+
```

Comment:
So we always need to do this additional lookup, its definitely more cleaner and simpler to use and I cannot think of how to avoid it with the new scheme.